pkg/clustermesh/remote_cluster.go

```diff
@@ -0,0 +1,185 @@
+
+	// ipCacheWatcher is the watcher that notifies about IP<->identity
+	// changes in the remote cluster
+	ipCacheWatcher *ipcache.IPIdentityWatcher
+
+	remoteIdentityCache *allocator.RemoteCache
+
+	// backend is the kvstor backend being used
+	backend kvstore.BackendOperations
+}
+
+var (
+	// skipKvstoreConnection skips the etcd connection, used for testing
+	skipKvstoreConnection bool
+)
+
+func (rc *remoteCluster) getLogger() *logrus.Entry {
+	return log.WithFields(logrus.Fields{
+		"name":   rc.name,
+		"config": rc.configPath,
+	})
+}
+
+func (rc *remoteCluster) restartRemoteConnection() {
+	rc.controllers.UpdateController(rc.remoteConnectionControllerName,
+		controller.ControllerParams{
+			DoFunc: func() error {
+				backend, err := kvstore.NewClient("etcd", map[string]string{
+					"etcd.config": rc.configPath,
+				})
+				if err != nil {
+					return err
+				}
+
+				remoteNodes, err := store.JoinSharedStore(store.Configuration{
+					Prefix:                  path.Join(node.NodeStorePrefix, rc.name),
+					KeyCreator:              rc.mesh.conf.NodeKeyCreator,
+					SynchronizationInterval: time.Minute,
+					Backend:                 backend,
+				})
+				if err != nil {
+					backend.Close()
+					return err
+				}
+
+				ipCacheWatcher := ipcache.NewIPIdentityWatcher(backend)
+				go ipCacheWatcher.Watch()
+
+				remoteIdentityCache := identity.WatchRemoteIdentities(backend)
+
+				rc.mutex.Lock()
+				rc.remoteNodes = remoteNodes
+				rc.backend = backend
+				rc.ipCacheWatcher = ipCacheWatcher
+				rc.remoteIdentityCache = remoteIdentityCache
+				rc.mutex.Unlock()
+
+				rc.getLogger().Info("Established connection to remote etcd")
+
+				return nil
+			},
+			StopFunc: func() error {
+				rc.ipCacheWatcher.Close()
+
+				rc.mutex.Lock()
+				if rc.remoteNodes != nil {
+					rc.remoteNodes.Close()
+				}
+				if rc.backend != nil {
+					rc.backend.Close()
+				}
+				if rc.remoteIdentityCache != nil {
+					rc.remoteIdentityCache.Close()
+				}
+				rc.mutex.Unlock()
+
+				rc.getLogger().Info("All resources of remote cluster cleaned up")
+
+				return nil
+			},
+		},
+	)
+}
+
+func (rc *remoteCluster) onInsert() {
+	rc.getLogger().Info("New remote cluster discovered")
+
+	if skipKvstoreConnection {
+		return
+	}
+
+	rc.remoteConnectionControllerName = fmt.Sprintf("remote-etcd-%s", rc.name)
+	rc.restartRemoteConnection()
+
+	go func() {
+		for {
+			val := <-rc.changed
+			if val {
+				rc.getLogger().Info("etcd configuration has changed, re-creating connection")
+				rc.restartRemoteConnection()
+			} else {
+				rc.getLogger().Info("Closing connection to remote etcd")
+				return
+			}
+		}
+	}()
+}
+
+func (rc *remoteCluster) onRemove() {
+	rc.controllers.RemoveAll()
+	close(rc.changed)
+
+	rc.getLogger().Info("Remote cluster disconnected")
+}
+
+func (rc *remoteCluster) isReady() bool {
+	rc.mutex.RLock()
+	defer rc.mutex.RUnlock()
+
+	return rc.backend != nil && rc.remoteNodes != nil && rc.ipCacheWatcher != nil
```

Comment:
When are any of these nil?

Reply:
These are nil while the controller is still connecting to the remote etcd and while the caches are being created. It is used from the tests to wait until all caches are created